lib/knetwork/kdns/records.go

```diff
@@ -0,0 +1,203 @@
+package kdns
+
+import (
+	"errors"
+	"fmt"
+	"github.com/miekg/dns"
+	"strings"
+)
+
+type RecordControllerErr error
+
+var (
+	UnSupportedTypeErr = errors.New("type of record not supported")
+)
+
+type RecordOp uint16
+
+var (
+	RecordDelete     RecordOp = 1
+	RecordWrite      RecordOp = 2
+	RecordWriteForce RecordOp = 3
+	RecordEdit       RecordOp = 4
+)
+
+// RecordController is a controller that specifically controls a single domain. It handles adding, removing and editing
+// dns records in place and should never error out. All methods write errors to the controllers error channel, and operate on
+// a fire and forget methodology.
+type RecordController struct {
+	ErrorChan chan RecordControllerErr
+
+	aRecords   chan []dns.RR
+	aRecordOps chan recordOperation
+
+	txtRecords   chan []dns.RR
+	txtRecordOps chan recordOperation
+}
+
+type recordOperation struct {
+	Type     RecordOp
+	Record   dns.RR
+	Records  []dns.RR
+	Keywords []string
+}
+
+// start will spin up the controller and begin handling dns data requests. It is non blocking and is called on NewRecordController
+func (rc *RecordController) start() {
+	go rc.watchARecords()
+	go rc.watchTxtRecords()
+}
+
+func handleOperation(src []dns.RR, operation recordOperation) []dns.RR {
+	switch operation.Type {
+	case RecordWrite:
+		return append(src, operation.Record)
+	case RecordEdit:
+		return append(DeleteIfContains(operation.Keywords, src), operation.Records...)
+	case RecordDelete:
+		return DeleteIfContains(operation.Keywords, src)
+	case RecordWriteForce:
+		return operation.Records
+	}
+	return src
+}
+
+func (rc *RecordController) watchARecords() {
```

Comment:
Added a close defer and a Stop method just in case, but unbuffered channels are actually part of garbage collection. 

`You needn't close every channel when you've finished with it. It's only necessary to close a channel when it is important to tell the receiving goroutines that all data have been sent. A channel that the garbage collector determines to be unreachable will have its resources reclaimed whether or not it is closed. (Don't confuse this with the close operation for open files. It is important to call the Close method on every file when you've finished with it.)`

Reply:
don't see this solved?